src/common/service/NetRemoteDataStreamingService.cxx

```diff
@@ -12,5 +12,7 @@ NetRemoteDataStreamingService::DataStreamUpload([[maybe_unused]] grpc::CallbackS
 {
     const NetRemoteApiTrace traceMe{};
 
-    return new Reactors::DataStreamReader(result);
+    m_dataStreamReaders.push_back(std::make_unique<Reactors::DataStreamReader>(result));
```

Comment:
It looks like this memory is only freed when the `NetRemoteDataStreamingService` instance is destroyed, and that means these will accumulate for the lifetime of the server.

I expected the service would be notified in some way that the RPC was complete and could then release its own reference on the reactor but reading the (embarrassingly sparse) documentation and examples, it seems this isn't the case. I'm surprised that gRPC exposes such a brittle design.

In the end, if we don't see a reason `NetRemoteDataStreamingService` needs to account for in-flight RPCs, then your original code (with `delete this` in the `OnDone()` function), should probably be used instead.

Reply:
The service essentially completes the RPC when it calls `Finish` inside the reactor class, which leads to the `OnDone` callback being invoked to allow for any cleanup. This was why I had `delete this` in that callback

Reply:
I could see why `delete this` was used, but that kind of usage makes me twitchy because we have to trust the calling code is going to respect that the memory is freed after `OnDone`. Hence I thought we could make it safer by using ref-counting, but I hadn't looked at the gRPC API holistically, and erroneously assumed there was some communication to the service about RPC completion status.

These kinds of implicit contracts, especially when undocumented, could change and/or be abused at any time and we'd have no real way of preventing it from happening, possibly causing very bad stuff to happen (ie. silent memory corruption due to accessing freed memory).